spec/factories/entitlement.rb

```diff
@@ -17,15 +17,15 @@
     name { code.humanize }
 
     trait :in_isolated_environment do
-      environment { build(:environment, :isolated, account:) }
+      environment { create(:environment, :isolated, account:) }
```

Comment:
Due to the way Rails v7.2 handles collection association inverses, the environment needs to be fully created before the other records are created, otherwise when the environment is saved, it also attempts to save any new associated records and that explodes because the environment is nil on those records.

E.g. given these records:

```ruby
environment = Environment.new
product     = Product.new(environment:)
```

The following returns a collection including the initialized product:

```ruby
environment.products # => [#<Product id=nil>]
```

In addition, when the environment is saved, so are all of its implicit associations:

```ruby
environment.save

environment.persisted? # => true
product.persisted?     # => true
```

Creating the environment first avoids this issue.

Reply:
For reference, look at `config.active_record.automatically_invert_plural_associations` and `config.active_record.has_many_inversing`, both of which default to `true` in Rails v7.2

Reply:
I think this may be a bug: https://github.com/rails/rails/issues/51853